### ISSUE TITLE
Android Pull Replication not honoring channels

### ISSUE THREAD

**Issue by pradipvaghasiya** [couchbase/couchbase-lite-android#887]
I am using 1.2.1 Couchbase Lite Android with forestDB and Sync Gateway 1.2

Please find below problem scenario. 
I am working on documents under Guest accessible channel
1. I start the one shot pull replication using channel id by setting setChannels. 
2. Replication completes and I get the document I am looking for. All is well upto now.
3. I call the same function to start same replication again with same channel name.
4. This time Pull replication downloads all guest documents which I have access to and does not respect the channel property which is set.

Function snippet

``` java
if (mReplication == null){ return false;}

        mReplication.setChannels(channels);
        if (mReplication.isPull() && (channels == null || channels.size() == 0) ){
            throw new RuntimeException("Pull Replication must have some channels");
        }
        System.out.println("Replication Channels: " + mReplication.getChannels());
        mListener = listener;
        Log.d(Constants.DEBUG_TAG_REPLICATION,"CouchOneShotReplicationType Status (Before Starting)" + mReplication.getStatus().toString()+ " Running: "+ mReplication.isRunning());

        if (mReplication.isRunning()) {
            mReplication.restart();
        } else {
            mReplication.start();
        }
        return true;
```

Application Logs: 
First Run Logs:

```
Replication Channels: [-far80c0rxe018dly3xt7js]
CouchOneShotReplicationType Status (Before Starting)REPLICATION_STOPPED Running: false
PullerInternal{unknown, pull, 67782}: Received invalid doc ID from _changes: {seq=3, id=_user/GUEST, changes=[]}
ation: Replication Send Success
{_id=-Far80c0RxE018dly3xt7jS, ....}
FollowEventManager Event ID replication Completed
```

Next Run Logs:

```
Replication Channels: [-far80c0rxe018dly3xt7js]
CouchOneShotReplicationType Status (Before Starting)REPLICATION_STOPPED Running: false
PullerInternal{unknown, pull, 922f5}: Received invalid doc ID from _changes: {seq=3, id=_user/GUEST, changes=[]}
ation: Replication Send Success
{_rev=2-d0354f332eb0fd2a5b17014f17cd1303, _id=-5W6LxhHd1VG9Lns9lo2QBG, ...}
{_rev=2-4bdb212ad4c6add67bd791723ba598c7, _id=-Far80c0RxE018dly3xt7jS, ...}
{_rev=2-3fd31d07148b0cd04845419ed7c6064e, _id=-FbmCppeNfpjmLlucSH-vSk, ...}
{_rev=2-58c4730998978800cc41105cfc9ba0d1, _id=-Lb3toHYlkWubVzqftbWsxd, ...}
FollowEventManager Event ID replication Completed
```

**NOTE: Same code on iOS for same Sync Gateway works just fine**

**Comment by pradipvaghasiya** [comment]
Update: 
This problem happens when I use same Replication instance to pull again. If every time I create new Replication and use setChannels then this does not happen.

**Comment by hideki** [comment]
Hi @pradipvaghasiya,
Thank you very much for bug report. It seem there is bug in re-using single-shot replicator. 
Thanks,

**Comment by pradipvaghasiya** [comment]
Hi @hideki,
The same issue is there with continous replication as well.

**Comment by mikemliu** [comment]
This is critical for applications that use a single continuous replication while handling cookie session renewals. Would be great if fix can be included in 1.3.1

**Comment by hideki** [comment]
Note to Self:

`Replication.setChannels(...)` internaly calls `replicationInternal.setChannels(channels);`
https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/Replication.java#L860-L862

In case of re-using Replicator, `replicationInternal` variable is recreated if replicator state is STOPPED when calling `start()`
https://github.com/couchbase/couchbase-lite-java-core/blob/master/src/main/java/com/couchbase/lite/replicator/Replication.java#L200-L225

As result, `channels` data is discarded.
